src/ux/table.ts

```diff
@@ -0,0 +1,321 @@
+      const sortKeys = this.filterColumnsFromHeaders(sortHeaders).map(
+        (c) =>
+          (v: Record<string, unknown>): unknown =>
+            v[c.key]
+      );
+      const sortKeysOrder = sorters.map((k) => (k.startsWith('-') ? 'desc' : 'asc'));
+      rows = orderBy(rows, sortKeys, sortKeysOrder);
+    }
+
+    // and filter columns
+    if (this.options.columns) {
+      const filters = this.options.columns.split(',');
+      this.columns = this.filterColumnsFromHeaders(filters);
+    } else if (!this.options.extended) {
+      // show extended columns/properties
+      this.columns = this.columns.filter((c) => !c.extended);
+    }
+
+    this.data = rows;
+  }
+
+  public display(): void {
+    const { data, options } = this;
+    // column truncation
+    //
+    // find max width for each column
+    const columns = this.columns.map((c) => {
+      const maxWidth = Math.max(sw('.'.padEnd(c.minWidth - 1)), sw(c.header), getWidestColumnWith(data, c.key)) + 1;
+      return {
+        ...c,
+        maxWidth,
+        width: maxWidth,
+      };
+    });
+
+    // terminal width
+    const maxWidth = stdtermwidth - 2;
+    // truncation logic
+    const shouldShorten = (): void => {
+      // don't shorten if full mode
+      if (options['no-truncate'] ?? (!process.stdout.isTTY && !process.env.CLI_UX_SKIP_TTY_CHECK)) return;
+
+      // don't shorten if there is enough screen width
+      const dataMaxWidth = sumBy(columns, (c) => c.width);
+      const overWidth = dataMaxWidth - maxWidth;
+      if (overWidth <= 0) return;
+
+      // not enough room, short all columns to minWidth
+      for (const col of columns) {
+        col.width = col.minWidth;
+      }
+
+      // if sum(minWidth's) is greater than term width
+      // nothing can be done so
+      // display all as minWidth
+      const dataMinWidth = sumBy(columns, (c) => c.minWidth);
+      if (dataMinWidth >= maxWidth) return;
+
+      // some wiggle room left, add it back to "needy" columns
+      let wiggleRoom = maxWidth - dataMinWidth;
+      const needyCols = columns
+        .map((c) => ({ key: c.key, needs: c.maxWidth - c.width }))
+        .sort((a, b) => a.needs - b.needs);
+      for (const { key, needs } of needyCols) {
+        if (!needs) continue;
+        const col = columns.find((c) => key === c.key);
+        if (!col) continue;
+        if (wiggleRoom > needs) {
+          col.width = col.width + needs;
+          wiggleRoom -= needs;
+        } else if (wiggleRoom) {
+          col.width = col.width + wiggleRoom;
+          wiggleRoom = 0;
+        }
+      }
+    };
+
+    shouldShorten();
+
+    // print table title
+    if (options.title) {
+      options.printLine(options.title);
+      // print title divider
+      options.printLine(
+        ''.padEnd(
+          columns.reduce((sum, col) => sum + col.width, 1),
+          '='
+        )
+      );
+
+      options.rowStart = '| ';
```

Comment:
options was read off of `this` and mutated here.  The consumer might not expect their table options to have been mutated (ex: I define some options that I want to use for several tables)

Reply:
How serious is this in your opinion? Given the complexity of this code, I'd feel more comfortable leaving it as it is for now

Reply:
ok, at least make a note of it for when I do the no-class rewrite 🤪